File: content/develop/use-cases/semantic-cache/redis-py/cache.py
"""
Redis semantic-cache helper backed by Redis Search.

Each cache entry lives as a Hash document at ``cache:<id>``. The hash
stores the user's prompt and the corresponding LLM response alongside
the raw float32 bytes of the prompt's 384-dimensional embedding and a
small set of metadata fields — tenant, locale, model version, and a
safety flag.

A single Redis Search index covers the embedding plus every metadata
field, so one ``FT.SEARCH`` call does an approximate-nearest-neighbour
lookup against the cached prompts with a TAG pre-filter applied in the
same pass — no cross-store joins, no extra round trips, and tenant
isolation is enforced *inside* the query rather than after the fact in
application code.

The lookup is thresholded: ``FT.SEARCH`` always returns the closest
cached prompt, but the cache only serves it as a hit when the cosine
distance is at or below ``distance_threshold``. Anything further away
is treated as a miss; the caller is expected to run the underlying LLM
and write the new prompt, response, and embedding back with ``put``.

Each cache entry is written with ``EXPIRE``, so stale answers age out
without manual cleanup; setting an eviction policy on the database
(``allkeys-lfu`` is the common choice) caps memory under pressure.
This helper expects a ``redis.Redis`` client constructed with
``decode_responses=False`` because the embedding field is binary;
mixing UTF-8 decoding into a binary path corrupts vectors. Text
fields are decoded explicitly where needed.
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass
from typing import Optional

import numpy as np
import redis
from redis.commands.search.field import (
    NumericField,
    TagField,
    TextField,
    VectorField,
)
from redis.commands.search.index_definition import IndexDefinition, IndexType
from redis.commands.search.query import Query


VECTOR_DIM_DEFAULT = 384


@dataclass
class CacheHit:
    """A cache lookup that returned a cached response.

    ``distance`` is the cosine distance ``FT.SEARCH`` reported for the
    nearest cached prompt (0 = identical, 2 = opposite). It is always
    at or below the threshold the lookup was run with.
    """

    id: str
    prompt: str
    response: str
    tenant: str
    locale: str
    model_version: str
    distance: float
    ttl_seconds: int
    hit_count: int

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "prompt": self.prompt,
            "response": self.response,
            "tenant": self.tenant,
            "locale": self.locale,
            "model_version": self.model_version,
            "distance": round(self.distance, 4),
            "ttl_seconds": self.ttl_seconds,
            "hit_count": self.hit_count,
        }


@dataclass
class CacheMiss:
    """A cache lookup that did not return a usable response.

    ``nearest_distance`` is the cosine distance to the closest cached
    prompt that *did* match the metadata filters. It is ``None`` if the
    cache had no entry in scope at all, which is what the demo UI
    shows as "no candidate" vs. "candidate too far".
    """

    nearest_distance: Optional[float]
    nearest_id: Optional[str]

    def to_dict(self) -> dict:
        return {
            "nearest_distance":
                round(self.nearest_distance, 4)
                if self.nearest_distance is not None else None,
            "nearest_id": self.nearest_id,
        }


class RedisSemanticCache:
    """Index, look up, and write a semantic cache of LLM responses."""

    def __init__(
        self,
        redis_client: Optional[redis.Redis] = None,
        index_name: str = "semcache:idx",
        key_prefix: str = "cache:",
        vector_dim: int = VECTOR_DIM_DEFAULT,
        distance_threshold: float = 0.5,
        default_ttl_seconds: int = 3600,
    ) -> None:
        self.redis = redis_client or redis.Redis(
            host="localhost",
            port=6379,
            decode_responses=False,
        )
        self.index_name = index_name
        self.key_prefix = key_prefix
        self.vector_dim = vector_dim
        self.distance_threshold = distance_threshold
        self.default_ttl_seconds = default_ttl_seconds

    # ------------------------------------------------------------------
    # Keys
    # ------------------------------------------------------------------

    def entry_key(self, entry_id: str) -> str:
        return f"{self.key_prefix}{entry_id}"

    # ------------------------------------------------------------------
    # Index management
    # ------------------------------------------------------------------

    def create_index(self) -> None:
        """Create the Redis Search index if it doesn't already exist.

        One index covers the embedding plus every metadata field, so a
        single ``FT.SEARCH`` can pre-filter by tenant / locale / model
        and then KNN-rank the matching documents in one pass. The
        ``prompt`` and ``response`` fields are stored as TEXT so the
        admin tooling can grep the cache by content, but the cache
        lookup itself is vector-only.
        """
        schema = (
            TextField("prompt"),
            TextField("response"),
            TagField("tenant"),
            TagField("locale"),
            TagField("model_version"),
            TagField("safety"),
            NumericField("created_ts", sortable=True),
            NumericField("hit_count", sortable=True),
            VectorField(
                "embedding",
                "HNSW",
                {
                    "TYPE": "FLOAT32",
                    "DIM": self.vector_dim,
                    "DISTANCE_METRIC": "COSINE",
                },
            ),
        )
        definition = IndexDefinition(
            prefix=[self.key_prefix], index_type=IndexType.HASH,
        )
        try:
            self.redis.ft(self.index_name).create_index(
                fields=schema, definition=definition,
            )
        except redis.ResponseError as exc:
            if "Index already exists" not in str(exc):
                raise

    def drop_index(self, delete_documents: bool = False) -> None:
        """Drop the search index. Optionally also delete cached entries."""
        try:
            self.redis.ft(self.index_name).dropindex(
                delete_documents=delete_documents,
            )
        except redis.ResponseError as exc:
            message = str(exc).lower()
            if "no such index" not in message \
                    and "unknown index name" not in message:
                raise

    # ------------------------------------------------------------------
    # Lookup
    # ------------------------------------------------------------------

    def lookup(
        self,
        query_vec: np.ndarray,
        tenant: str | None = None,
        locale: str | None = None,
        model_version: str | None = None,
        safety: str | None = "ok",
        distance_threshold: float | None = None,
    ) -> CacheHit | CacheMiss:
        """Find the nearest in-scope cached prompt and decide hit / miss.

        ``FT.SEARCH`` returns the single nearest entry that satisfies
        the TAG pre-filters. The lookup is a hit only if the reported
        cosine distance is at or below ``distance_threshold`` (or the
        instance default). Anything further away is a miss with the
        candidate distance attached so the caller can log it.

        On a hit, the entry's ``hit_count`` is incremented atomically
        with ``HINCRBY`` so the demo UI can show which entries are
        load-bearing. The TTL is refreshed on every hit so frequently
        used answers don't age out under cold tail entries.
        """
        threshold = (
            distance_threshold if distance_threshold is not None
            else self.distance_threshold
        )

        # Match the shape check that ``put`` performs. A wrong-dim
        # vector would otherwise hit Redis as a malformed FT.SEARCH
        # parameter and surface as a server-side parse error instead
        # of a clear caller-side ValueError.
        if query_vec.shape != (self.vector_dim,):
            raise ValueError(
                f"query_vec has shape {query_vec.shape}; "
                f"index expects ({self.vector_dim},)"
            )

        filter_clause = self._build_filter_clause(
            tenant=tenant,
            locale=locale,
            model_version=model_version,
            safety=safety,
        )
        knn_query = f"{filter_clause}=>[KNN 1 @embedding $vec AS distance]"
        q = (
            Query(knn_query)
            .sort_by("distance")
            .return_fields(
                "prompt", "response", "tenant", "locale",
                "model_version", "hit_count", "distance",
            )
            .paging(0, 1)
            .dialect(2)
        )
        result = self.redis.ft(self.index_name).search(
            q, query_params={"vec": query_vec.astype(np.float32).tobytes()},
        )

        if not result.docs:
            return CacheMiss(nearest_distance=None, nearest_id=None)

        doc = result.docs[0]
        raw_key = _decode(getattr(doc, "id", ""))
        entry_id = (
            raw_key[len(self.key_prefix):] if raw_key.startswith(self.key_prefix)
            else raw_key
        )
        distance = float(_decode(getattr(doc, "distance", "0")) or 0)

        if distance > threshold:
            return CacheMiss(nearest_distance=distance, nearest_id=entry_id)

        # The hash may have expired between FT.SEARCH returning the
        # row and us getting here — the search index lags expirations
        # by its periodic scan. If we just blindly ``HINCRBY``-ed,
        # Redis would helpfully recreate the hash with only
        # ``hit_count`` set and the search index would then log it as
        # an indexing failure (no embedding, no metadata). EXISTS
        # narrows that race to the pipeline round-trip; a strictly
        # race-free version would wrap the bump in a Lua script that
        # checks existence and acts in one server-side step.
        entry_key = self.entry_key(entry_id)
        if not self.redis.exists(entry_key):
            return CacheMiss(nearest_distance=distance, nearest_id=entry_id)

        # MULTI/EXEC the three writes so they apply as a unit on the
        # server — a partial failure between HINCRBY and EXPIRE would
        # otherwise leave the entry without a refreshed TTL.
        pipe = self.redis.pipeline(transaction=True)
        pipe.hincrby(entry_key, "hit_count", 1)
        pipe.expire(entry_key, self.default_ttl_seconds)
        pipe.ttl(entry_key)
        new_hit_count, _expired, ttl = pipe.execute()

        return CacheHit(
            id=entry_id,
            prompt=_decode(getattr(doc, "prompt", "")),
            response=_decode(getattr(doc, "response", "")),
            tenant=_decode(getattr(doc, "tenant", "")),
            locale=_decode(getattr(doc, "locale", "")),
            model_version=_decode(getattr(doc, "model_version", "")),
            distance=distance,
            ttl_seconds=int(ttl) if ttl and ttl > 0 else self.default_ttl_seconds,
            hit_count=int(new_hit_count),
        )

    # ------------------------------------------------------------------
    # Write
    # ------------------------------------------------------------------

    def put(
        self,
        prompt: str,
        response: str,
        embedding: np.ndarray,
        tenant: str = "default",
        locale: str = "en",
        model_version: str = "gpt-4.5-2026",
        safety: str = "ok",
        ttl_seconds: int | None = None,
        entry_id: str | None = None,
    ) -> str:
        """Write a new cache entry and return its id.

        The embedding is stored as raw little-endian float32 bytes —
        the encoding Redis Search expects from a FLOAT32 vector field.
        ``EXPIRE`` on the key gives every entry a bounded lifetime;
        combine with an ``allkeys-lfu`` eviction policy on the database
        to cap memory under pressure too.
        """
        if embedding.shape != (self.vector_dim,):
            raise ValueError(
                f"embedding has shape {embedding.shape}; "
                f"index expects ({self.vector_dim},)"
            )
        if embedding.dtype != np.float32:
            embedding = embedding.astype(np.float32, copy=False)

        entry_id = entry_id or uuid.uuid4().hex[:12]
        key = self.entry_key(entry_id)
        ttl = ttl_seconds if ttl_seconds is not None else self.default_ttl_seconds

        mapping = {
            "prompt": prompt,
            "response": response,
            "tenant": tenant,
            "locale": locale,
            "model_version": model_version,
            "safety": safety,
            "created_ts": str(time.time()),
            "hit_count": "0",
            "embedding": embedding.tobytes(),
        }
        # MULTI/EXEC so HSET and EXPIRE either both apply or neither
        # does. Without the transaction wrapper a connection drop
        # between the two writes could leave the entry without a TTL
        # and the cache would then keep an answer past its intended
        # lifetime (or forever, on a database with no eviction policy).
        pipe = self.redis.pipeline(transaction=True)
        pipe.hset(key, mapping=mapping)
        pipe.expire(key, ttl)
        pipe.execute()
        return entry_id

    # ------------------------------------------------------------------
    # Filter clause
    # ------------------------------------------------------------------

    # Characters Redis Search treats as syntax inside a TAG value; any
    # of them in a user-supplied filter must be backslash-escaped or
    # the surrounding ``{...}`` block won't parse correctly.
    _TAG_SPECIAL = set("\\,.<>{}[]\"':;!@#$%^&*()-+=~| ")

    @classmethod
    def _escape_tag_value(cls, value: str) -> str:
        return "".join(
            "\\" + ch if ch in cls._TAG_SPECIAL else ch for ch in value
        )

    @classmethod
    def _build_filter_clause(
        cls,
        *,
        tenant: str | None,
        locale: str | None,
        model_version: str | None,
        safety: str | None,
    ) -> str:
        clauses: list[str] = []
        if tenant:
            clauses.append(f"@tenant:{{{cls._escape_tag_value(tenant)}}}")
        if locale:
            clauses.append(f"@locale:{{{cls._escape_tag_value(locale)}}}")
        if model_version:
            clauses.append(
                f"@model_version:{{{cls._escape_tag_value(model_version)}}}"
            )
        if safety:
            clauses.append(f"@safety:{{{cls._escape_tag_value(safety)}}}")
        return "(" + " ".join(clauses) + ")" if clauses else "(*)"

    # ------------------------------------------------------------------
    # Inspection / admin
    # ------------------------------------------------------------------

    def index_info(self) -> dict:
        """Subset of ``FT.INFO`` useful for the demo UI."""
        try:
            info = self.redis.ft(self.index_name).info()
        except redis.ResponseError:
            return {"num_docs": 0, "indexing_failures": 0,
                    "vector_index_size_mb": 0.0}
        return {
            "num_docs": int(info.get("num_docs", 0)),
            "indexing_failures":
                int(info.get("hash_indexing_failures", 0)),
            "vector_index_size_mb": _safe_mb(info),
        }

    def list_entries(self, limit: int = 100) -> list[dict]:
        """Return every cached entry (no embedding) for the admin UI."""
        q = (
            Query("*")
            .return_fields(
                "prompt", "response", "tenant", "locale",
                "model_version", "safety", "created_ts", "hit_count",
            )
            .paging(0, limit)
            .sort_by("created_ts", asc=False)
        )
        result = self.redis.ft(self.index_name).search(q)
        out: list[dict] = []
        for doc in result.docs:
            raw_key = _decode(getattr(doc, "id", ""))
            entry_id = (
                raw_key[len(self.key_prefix):]
                if raw_key.startswith(self.key_prefix) else raw_key
            )
            ttl = self.redis.ttl(self.entry_key(entry_id))
            out.append({
                "id": entry_id,
                "prompt": _decode(getattr(doc, "prompt", "")),
                "response": _decode(getattr(doc, "response", "")),
                "tenant": _decode(getattr(doc, "tenant", "")),
                "locale": _decode(getattr(doc, "locale", "")),
                "model_version": _decode(getattr(doc, "model_version", "")),
                "safety": _decode(getattr(doc, "safety", "")),
                "hit_count":
                    int(_decode(getattr(doc, "hit_count", "0")) or 0),
                "ttl_seconds": int(ttl) if ttl and ttl > 0 else 0,
                "created_ts":
                    float(_decode(getattr(doc, "created_ts", "0")) or 0),
            })
        return out

    def delete_entry(self, entry_id: str) -> bool:
        """Drop a single entry. Returns ``True`` if the key existed."""
        return bool(self.redis.delete(self.entry_key(entry_id)))

    def clear(self) -> int:
        """Drop the index and every cached entry.

        Returns the number of entries that were removed. Used by the
        demo's "reset" button — in production the equivalent is just
        ``FLUSHDB`` on a dedicated cache database, or letting TTLs
        expire naturally.
        """
        before = self.index_info()["num_docs"]
        self.drop_index(delete_documents=True)
        self.create_index()
        return before


def _decode(value) -> str:
    if value is None:
        return ""
    if isinstance(value, bytes):
        return value.decode("utf-8")
    return value


def _safe_mb(info: dict) -> float:
    try:
        return float(info.get("vector_index_sz_mb", 0.0) or 0.0)
    except (TypeError, ValueError):
        return 0.0

File: content/develop/use-cases/semantic-cache/ruby/lib/cache.rb
# Redis semantic-cache helper backed by Redis Search (Ruby).
#
# Each cache entry lives as a Hash document at `cache:<id>`. The hash
# stores the user's prompt and the corresponding LLM response alongside
# the raw float32 bytes of the prompt's 384-dimensional embedding and a
# small set of metadata fields — tenant, locale, model version, and a
# safety flag.
#
# A single Redis Search index covers the embedding plus every metadata
# field, so one `FT.SEARCH` call does an approximate-nearest-neighbour
# lookup against the cached prompts with a TAG pre-filter applied in
# the same pass — no cross-store joins, no extra round trips, and
# tenant isolation is enforced *inside* the query rather than after
# the fact in application code.
#
# The lookup is thresholded: `FT.SEARCH` always returns the closest
# cached prompt, but the cache only serves it as a hit when the cosine
# distance is at or below `distance_threshold`. Anything further away
# is treated as a miss; the caller is expected to run the underlying
# LLM and write the new prompt, response, and embedding back with
# `put`.
#
# Each cache entry is written with `EXPIRE`, so stale answers age out
# without manual cleanup; setting an eviction policy on the database
# (`allkeys-lfu` is the common choice) caps memory under pressure.
# This helper assumes the `redis-rb` client default — String I/O — and
# packs the binary embedding bytes into a String with ASCII-8BIT
# encoding so the protocol writer transmits the exact bytes without
# any UTF-8 transcoding.

require 'redis'
require 'securerandom'
require 'set'

module SemCache
  VECTOR_DIM_DEFAULT = 384

  # A cache lookup that returned a cached response. `distance` is the
  # cosine distance `FT.SEARCH` reported for the nearest cached
  # prompt (0 = identical, 2 = opposite). It is always at or below
  # the threshold the lookup was run with.
  CacheHit = Struct.new(
    :id, :prompt, :response, :tenant, :locale, :model_version,
    :distance, :ttl_seconds, :hit_count, keyword_init: true
  ) do
    def to_h
      {
        id: id,
        prompt: prompt,
        response: response,
        tenant: tenant,
        locale: locale,
        model_version: model_version,
        distance: distance.round(4),
        ttl_seconds: ttl_seconds,
        hit_count: hit_count
      }
    end

    def hit?
      true
    end
  end

  # A cache lookup that did not return a usable response.
  # `nearest_distance` is the cosine distance to the closest cached
  # prompt that *did* match the metadata filters. It is `nil` if the
  # cache had no entry in scope at all, which is what the demo UI
  # shows as "no candidate" vs. "candidate too far".
  CacheMiss = Struct.new(
    :nearest_distance, :nearest_id, keyword_init: true
  ) do
    def to_h
      {
        nearest_distance: nearest_distance ? nearest_distance.round(4) : nil,
        nearest_id: nearest_id
      }
    end

    def hit?
      false
    end
  end

  class RedisSemanticCache
    # Characters Redis Search treats as syntax inside a TAG value; any
    # of them in a user-supplied filter must be backslash-escaped or
    # the surrounding `{...}` block won't parse correctly.
    TAG_SPECIAL = Set.new("\\,.<>{}[]\"':;!@#$%^&*()-+=~| ".chars).freeze

    attr_reader :redis, :index_name, :key_prefix, :vector_dim,
                :distance_threshold, :default_ttl_seconds

    def initialize(redis_client: nil,
                   index_name: 'semcache:idx',
                   key_prefix: 'cache:',
                   vector_dim: VECTOR_DIM_DEFAULT,
                   distance_threshold: 0.5,
                   default_ttl_seconds: 3600)
      @redis = redis_client || Redis.new(host: 'localhost', port: 6379)
      @index_name = index_name
      @key_prefix = key_prefix
      @vector_dim = vector_dim
      @distance_threshold = distance_threshold
      @default_ttl_seconds = default_ttl_seconds
    end

    # ----------------------------------------------------------------
    # Keys
    # ----------------------------------------------------------------

    def entry_key(entry_id)
      "#{@key_prefix}#{entry_id}"
    end

    # ----------------------------------------------------------------
    # Index management
    # ----------------------------------------------------------------

    # Create the Redis Search index if it doesn't already exist. One
    # index covers the embedding plus every metadata field, so a
    # single `FT.SEARCH` can pre-filter by tenant / locale / model and
    # then KNN-rank the matching documents in one pass. The `prompt`
    # and `response` fields are stored as TEXT so the admin tooling
    # can grep the cache by content, but the cache lookup itself is
    # vector-only.
    def create_index
      args = [
        'FT.CREATE', @index_name,
        'ON', 'HASH',
        'PREFIX', '1', @key_prefix,
        'SCHEMA',
        'prompt',        'TEXT',
        'response',      'TEXT',
        'tenant',        'TAG',
        'locale',        'TAG',
        'model_version', 'TAG',
        'safety',        'TAG',
        'created_ts',    'NUMERIC', 'SORTABLE',
        'hit_count',     'NUMERIC', 'SORTABLE',
        'embedding',     'VECTOR', 'HNSW', '6',
        'TYPE', 'FLOAT32',
        'DIM', @vector_dim.to_s,
        'DISTANCE_METRIC', 'COSINE'
      ]
      @redis.call(*args)
    rescue Redis::CommandError => e
      raise unless e.message.include?('Index already exists')
    end

    # Drop the search index. Optionally also delete cached entries.
    def drop_index(delete_documents: false)
      args = ['FT.DROPINDEX', @index_name]
      args << 'DD' if delete_documents
      @redis.call(*args)
    rescue Redis::CommandError => e
      message = e.message.downcase
      raise unless message.include?('no such index') || message.include?('unknown index name')
    end

    # ----------------------------------------------------------------
    # Lookup
    # ----------------------------------------------------------------

    # Find the nearest in-scope cached prompt and decide hit / miss.
    #
    # `FT.SEARCH` returns the single nearest entry that satisfies the
    # TAG pre-filters. The lookup is a hit only if the reported cosine
    # distance is at or below `distance_threshold` (or the instance
    # default). Anything further away is a miss with the candidate
    # distance attached so the caller can log it.
    #
    # On a hit, the entry's `hit_count` is incremented atomically with
    # `HINCRBY` so the demo UI can show which entries are
    # load-bearing. The TTL is refreshed on every hit so frequently
    # used answers don't age out under cold tail entries.
    def lookup(query_vec, tenant: nil, locale: nil, model_version: nil,
               safety: 'ok', distance_threshold: nil)
      validate_dim!(query_vec, 'query_vec')

      threshold = distance_threshold || @distance_threshold

      filter_clause = self.class.build_filter_clause(
        tenant: tenant, locale: locale,
        model_version: model_version, safety: safety
      )
      query_str = "#{filter_clause}=>[KNN 1 @embedding $vec AS distance]"
      vec_bytes = LocalEmbedder.to_bytes(query_vec)

      args = [
        'FT.SEARCH', @index_name, query_str,
        'PARAMS', '2', 'vec', vec_bytes,
        'SORTBY', 'distance', 'ASC',
        'RETURN', '7',
        'prompt', 'response', 'tenant', 'locale',
        'model_version', 'hit_count', 'distance',
        'LIMIT', '0', '1',
        'DIALECT', '2'
      ]
      result = @redis.call(*args)
      docs = parse_search_result(result)

      return CacheMiss.new(nearest_distance: nil, nearest_id: nil) if docs.empty?

      doc = docs.first
      raw_key = doc[:_key]
      entry_id = raw_key.start_with?(@key_prefix) ? raw_key[@key_prefix.length..] : raw_key
      distance = doc[:distance].to_f

      if distance > threshold
        return CacheMiss.new(nearest_distance: distance, nearest_id: entry_id)
      end

      # The hash may have expired between FT.SEARCH returning the row
      # and us getting here — the search index lags expirations by its
      # periodic scan. If we just blindly HINCRBY-ed, Redis would
      # helpfully recreate the hash with only `hit_count` set and the
      # search index would then log it as an indexing failure (no
      # embedding, no metadata). EXISTS narrows that race to the
      # pipeline round-trip; a strictly race-free version would wrap
      # the bump in a Lua script that checks existence and acts in
      # one server-side step.
      ek = entry_key(entry_id)
      unless @redis.exists?(ek)
        return CacheMiss.new(nearest_distance: distance, nearest_id: entry_id)
      end

      # MULTI/EXEC the three writes so they apply as a unit on the
      # server — a partial failure between HINCRBY and EXPIRE would
      # otherwise leave the entry without a refreshed TTL.
      replies = @redis.multi do |m|
        m.hincrby(ek, 'hit_count', 1)
        m.expire(ek, @default_ttl_seconds)
        m.ttl(ek)
      end
      new_hit_count, _expired, ttl = replies

      CacheHit.new(
        id: entry_id,
        prompt: doc[:prompt] || '',
        response: doc[:response] || '',
        tenant: doc[:tenant] || '',
        locale: doc[:locale] || '',
        model_version: doc[:model_version] || '',
        distance: distance,
        ttl_seconds: ttl && ttl.positive? ? ttl.to_i : @default_ttl_seconds,
        hit_count: new_hit_count.to_i
      )
    end

    # ----------------------------------------------------------------
    # Write
    # ----------------------------------------------------------------

    # Write a new cache entry and return its id.
    #
    # The embedding is stored as raw little-endian float32 bytes — the
    # encoding Redis Search expects from a FLOAT32 vector field.
    # `EXPIRE` on the key gives every entry a bounded lifetime;
    # combine with an `allkeys-lfu` eviction policy on the database
    # to cap memory under pressure too.
    def put(prompt:, response:, embedding:,
            tenant: 'default', locale: 'en',
            model_version: 'gpt-4.5-2026', safety: 'ok',
            ttl_seconds: nil, entry_id: nil)
      validate_dim!(embedding, 'embedding')

      id = entry_id || SecureRandom.hex(6) # 12 hex chars, matches sibling demos
      key = entry_key(id)
      ttl = ttl_seconds || @default_ttl_seconds
      vec_bytes = LocalEmbedder.to_bytes(embedding)

      # MULTI/EXEC so HSET and EXPIRE either both apply or neither
      # does. Without the transaction wrapper a connection drop
      # between the two writes could leave the entry without a TTL
      # and the cache would then keep an answer past its intended
      # lifetime (or forever, on a database with no eviction policy).
      #
      # `redis-rb`'s HSET takes a flat list of field/value pairs or a
      # Hash. We pass the flat list so the binary `vec_bytes` is sent
      # as one argument without any String-to-Hash key coercion
      # touching it.
      @redis.multi do |m|
        m.hset(key,
               'prompt', prompt,
               'response', response,
               'tenant', tenant,
               'locale', locale,
               'model_version', model_version,
               'safety', safety,
               'created_ts', Time.now.to_f.to_s,
               'hit_count', '0',
               'embedding', vec_bytes)
        m.expire(key, ttl)
      end
      id
    end

    # ----------------------------------------------------------------
    # Filter clause
    # ----------------------------------------------------------------

    def self.escape_tag_value(value)
      value.each_char.map { |c| TAG_SPECIAL.include?(c) ? "\\#{c}" : c }.join
    end

    def self.build_filter_clause(tenant:, locale:, model_version:, safety:)
      clauses = []
      clauses << "@tenant:{#{escape_tag_value(tenant)}}" if tenant && !tenant.empty?
      clauses << "@locale:{#{escape_tag_value(locale)}}" if locale && !locale.empty?
      clauses << "@model_version:{#{escape_tag_value(model_version)}}" if model_version && !model_version.empty?
      clauses << "@safety:{#{escape_tag_value(safety)}}" if safety && !safety.empty?
      clauses.empty? ? '(*)' : "(#{clauses.join(' ')})"
    end

    # ----------------------------------------------------------------
    # Inspection / admin
    # ----------------------------------------------------------------

    # Subset of `FT.INFO` useful for the demo UI.
    def index_info
      raw = @redis.call('FT.INFO', @index_name)
      info = ft_info_to_hash(raw)
      {
        num_docs: (info['num_docs'] || 0).to_i,
        indexing_failures: (info['hash_indexing_failures'] || 0).to_i,
        vector_index_size_mb: (info['vector_index_sz_mb'] || 0).to_f
      }
    rescue Redis::CommandError
      { num_docs: 0, indexing_failures: 0, vector_index_size_mb: 0.0 }
    end

    # Return every cached entry (no embedding) for the admin UI.
    def list_entries(limit: 100)
      args = [
        'FT.SEARCH', @index_name, '*',
        'RETURN', '8',
        'prompt', 'response', 'tenant', 'locale',
        'model_version', 'safety', 'created_ts', 'hit_count',
        'LIMIT', '0', limit.to_s,
        'SORTBY', 'created_ts', 'DESC',
        'DIALECT', '2'
      ]
      result = @redis.call(*args)
      parse_search_result(result).map do |doc|
        raw_key = doc[:_key]
        entry_id = raw_key.start_with?(@key_prefix) ? raw_key[@key_prefix.length..] : raw_key
        ttl = @redis.ttl(entry_key(entry_id))
        {
          id: entry_id,
          prompt: doc[:prompt] || '',
          response: doc[:response] || '',
          tenant: doc[:tenant] || '',
          locale: doc[:locale] || '',
          model_version: doc[:model_version] || '',
          safety: doc[:safety] || '',
          hit_count: (doc[:hit_count] || '0').to_i,
          ttl_seconds: ttl && ttl.positive? ? ttl.to_i : 0,
          created_ts: (doc[:created_ts] || '0').to_f
        }
      end
    end

    # Drop a single entry. Returns true if the key existed.
    def delete_entry(entry_id)
      @redis.del(entry_key(entry_id)).positive?
    end

    # Drop the index and every cached entry. Returns the number of
    # entries that were removed. Used by the demo's "reset" button —
    # in production the equivalent is just `FLUSHDB` on a dedicated
    # cache database, or letting TTLs expire naturally.
    def clear
      before = index_info[:num_docs]
      drop_index(delete_documents: true)
      create_index
      before
    end

    # ----------------------------------------------------------------
    # Internals
    # ----------------------------------------------------------------

    private

    def validate_dim!(vector, label)
      unless vector.respond_to?(:length) && vector.length == @vector_dim
        actual = vector.respond_to?(:length) ? vector.length : 'unknown'
        raise ArgumentError,
              "#{label} has length #{actual}; index expects #{@vector_dim}"
      end
    end

    # Parse the raw `FT.SEARCH` reply (RESP2 layout). The shape is:
    #   [ total, key1, [field1, value1, field2, value2, ...], key2, [...], ... ]
    # where each key is followed by a flat field/value array.
    def parse_search_result(reply)
      return [] unless reply.is_a?(Array) && reply.length >= 1
      _total = reply[0]
      docs = []
      i = 1
      while i < reply.length
        key = reply[i]
        fields = reply[i + 1]
        i += 2
        next if fields.nil?
        doc = { _key: key }
        # `fields` is a flat [k, v, k, v, ...] array; convert pairs to
        # symbol-keyed entries on the doc hash for easy lookup.
        j = 0
        while j < fields.length
          field_name = fields[j].to_s
          field_value = fields[j + 1]
          doc[field_name.to_sym] = field_value
          j += 2
        end
        docs << doc
      end
      docs
    end

    # `FT.INFO` returns a flat alternating key/value array. Lift it to
    # a string-keyed hash, ignoring nested arrays we don't need.
    def ft_info_to_hash(reply)
      return {} unless reply.is_a?(Array)
      out = {}
      i = 0
      while i < reply.length
        out[reply[i].to_s] = reply[i + 1]
        i += 2
      end
      out
    end
  end
end
